resources/type-system/subtyping.md

```diff
@@ -123,66 +131,92 @@ with promoted type variables.
 
 ### Rules
 
+**Note the convention described in the section above**
+
 We say that a type `T0` is a subtype of a type `T1` (written `T0 <: T1`) when:
 
-- **Reflexivity**: `T0` and `T1` are the same type.
+- **Reflexivity**: if `T0` and `T1` are the same type then `T0 <: T1`
   - *Note that this check is necessary as the base case for primitive types, and
     type variables but not for composite types.  In particular, algorithmically
     a structural equality check is admissible, but not required
     here. Pragmatically, non-constant time identity checks here are
     counter-productive*
 
-- **Right Top**: `T1` is a top type (i.e. `Object`, `dynamic`, or `void`).
+- **Right Top**: if `T1` is a top type (i.e. `dynamic`, or `void`, or `Object?`)
+  then `T0 <: T1`
 
-- **Left Bottom**: `T0` is `Never`
+- **Left Bottom**: if `T0` is `Never` or `Null!` then `T0 <: T1`
 
-- **Right Legacy Top**: `T1` is a legacy top type (i.e. `Object*`, `dynamic*`,
-  or `void*`).
+- **Right Object**: if `T1` is `Object` then:
+    - if `T0` is an unpromoted type variable with bound `B` then `T0 <: T1` iff
+      `B <: Object`
+    - if `T0` is a promoted type variable `X & S` then `T0 <: T1` iff `S <:
+      Object`
+    - if `T0` is `FutureOr<S>` for some `S`, then `T0 <: T1` iff `S <: Object`.
+    - if `T0` is `Null`, `dynamic`, `void`, or `S?` for any `S`, then the
+      subtyping does not hold (per above, the result of the subtyping query is
+      false).
+    - Otherwise `T0 <: T1` is true.
```

Comment:
> However, we could stick to a positive model and I believe it wouldn't have worse performance:

By definition it has the same performance - we're specifying the same relation either way, so it admits the same implementations.  The only difference is that the implementation has to rediscover that there is a total order on the checks.

> Maybe it's just me, but I suspect that the ruleset as a whole will be a bit more readable if we try to make it order-independent, especially if it doesn't cost anything.

Well... I rather suspect the opposite.  But I'm happy to review an attempt to rewrite this in an order independent way.

Reply:
Drive-by comment: speaking as one of the people who is going to have to write the implementation, I prefer the order-dependent formulation, because it admits a straightforward direct translation to code.  So I'll have more confidence that the implementation and the spec match.